sdk/azidentity/internal/cache.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package internal
+
+import "github.com/AzureAD/microsoft-authentication-library-for-go/apps/cache"
+
+// Cache represents a persistent token cache
+type Cache struct {
+	// impl is embedded as a pointer so a Cache
+	// can carry persistent state across copies
+	*impl
+}
+
+// impl is a Cache's private implementation
+type impl struct {
+	// factory constructs implementations
+	factory func(bool) (cache.ExportReplace, error)
+	// cae and noCAE are previously constructed implementations
+	cae, noCAE cache.ExportReplace
+}
+
+func (i *impl) exportReplace(cae bool) (cache.ExportReplace, error) {
+	if i == nil {
+		// zero-value Cache: return a nil ExportReplace and MSAL will cache in memory
+		return nil, nil
+	}
+	var (
+		err error
+		xr  cache.ExportReplace
+	)
+	if cae {
+		if i.cae == nil {
+			if xr, err = i.factory(cae); err == nil {
+				i.cae = xr
```

Comment:
If I have multiple credentials sharing a cache, can `exportReplace` be called concurrently? Or will synchronization happen at a higher level? Wondering if this is going to race (also line 42).

Reply:
Good catch. A race is possible but probably not dangerous because a cache's name determines its storage and accessing that is synchronized elsewhere with a lock file. This looks bad and should trip the race detector though, I'll add synchronization.